File: src/AdventOfCode/Puzzles/Y2020/Day01.cs
﻿// Copyright (c) Martin Costello, 2015. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

namespace MartinCostello.AdventOfCode.Puzzles.Y2020
{
    using System.Collections.Generic;
    using System.Linq;

    /// <summary>
    /// A class representing the puzzle for <c>https://adventofcode.com/2020/day/1</c>. This class cannot be inherited.
    /// </summary>
    public sealed class Day01 : Puzzle2020
    {
        /// <summary>
        /// Gets the product of the two input values that sum to a value of 2020.
        /// </summary>
        public int ProductOf2020SumFrom2 { get; private set; }

        /// <summary>
        /// Gets the product of the three input values that sum to a value of 2020.
        /// </summary>
        public int ProductOf2020SumFrom3 { get; private set; }

        /// <summary>
        /// Gets the product of a number of values from the specified set of values
        /// that which when added together equal 2020.
        /// </summary>
        /// <param name="values">The values to find the 2020 sum's product from.</param>
        /// <param name="take">The number of values to use to reach the 2020 target.</param>
        /// <returns>
        /// The product of the specified number of values that sum to a value of 2020.
        /// </returns>
        public static int Get2020Product(IEnumerable<string> values, int take)
        {
            int[] expenses = values
                .Select((p) => ParseInt32(p))
                .ToArray();

            var result = Maths.GetPermutations(expenses, take);

            return result
                .Where((p) => p.Sum() == 2020)
                .Select((p) => p.Aggregate((x, y) => x * y))
                .First();
        }

        /// <inheritdoc />
        protected override int SolveCore(string[] args)
        {
            IList<string> values = ReadResourceAsLines();

            ProductOf2020SumFrom2 = Get2020Product(values, 2);
            ProductOf2020SumFrom3 = Get2020Product(values, 3);

            if (Verbose)
            {
                Logger.WriteLine("The product of the two entries that sum to 2020 is {0}.", ProductOf2020SumFrom2);
                Logger.WriteLine("The product of the three entries that sum to 2020 is {0}.", ProductOf2020SumFrom3);
            }

            return 0;
        }
    }
}

File: src/AdventOfCode.Site/scripts/ts/puzzles/2023/Day01.ts
// Copyright (c) Martin Costello, 2015. All rights reserved.
// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.

import { Puzzle2023 } from './Puzzle2023';

export class Day01 extends Puzzle2023 {
    solution1: number;
    solution2: number;

    override get solved() {
        return false;
    }

    override get name() {
        return '';
    }

    override get day() {
        return 1;
    }

    static solve(values: string[]): number {
        return -1;
    }

    override solveCore(_: string[]) {
        const values = this.readResourceAsLines();

        this.solution1 = Day01.solve(values);
        this.solution2 = Day01.solve(values);

        console.info(`${this.solution1}`);
        console.info(`${this.solution2}`);

        return this.createResult([this.solution1, this.solution2]);
    }
}
